### ISSUE TITLE
Syntax error caused by broken expression parsing in command substitution

### ISSUE THREAD

**Issue by lukaswrz** [oilshell/oil#1387]
If the following script is executed with Oil, it will output `true`.

```
if (true) {
  write true
}
```

If the above script is wrapped in a command substitution, Oil will produce a syntax error:

```
write $(
  if (true) {
    write true
  }
)
```

The error output is:

```
    if (true) {
            ^
test.oil:2: Syntax error in expression (near Id.Eof_RParen)
```

However, other constructs that involve parentheses (such as `case`) appear to work fine in command substitutions:

```
write $(
  case 'README.md' {
    (*.md)
      write 'This is a markdown file'
      ;;
  }
)
```

The above will (as expected) output `This is a markdown file`.

Tested on Oil 0.12.7 as well as an up-to-date dev build.

**Comment by lukaswrz** [comment]
BTW, the same issue happens with parentheses in `$[]`:

```
const r = $(
  write $[len('foo')]
)
```

But it doesn't throw a syntax error if no parentheses are involved, e.g.:

```
const r = $(
  write $[5 ** 3]
)
```

**Comment by andychu** [comment]
Hm thanks for the report!  This  does seem like  a bug, I will look into it

The `case` works because there is a stack of parens; maybe the expression parser is not properly using that stack

**Comment by andychu** [comment]
Thank you, just fixed this!  Will be out with the next release

Let me know if you see any other problems :-)

**Comment by lukaswrz** [comment]
Thanks, it looks like that bug is fixed now.

However, I tried all the examples that I gave, and it seems like there's another (perhaps unrelated?) problem:

```
const r = $(
  write $[5 ** 3]
)
```

Throws this error:

```
[ interactive ]:1: Invalid word while parsing command
```

Not exactly sure what is causing this. It happens both in interactive mode and when executed as a script. The problem goes away as soon as the newlines are removed.

It doesn't happen on release 0.13.1, which makes me think it could be related...

Should I open a new issue about this?

**Comment by andychu** [comment]
Hm interesting, let me look into what happened ... it could be a side effect of fixing that

(The underlying issue is that we are playing some tricks to interleave shell and Oil, and shell has a lot of meanings of `)` )

Thanks for testing it!

**Comment by andychu** [comment]
Yes that was an error in yesterday's fix  -- thank you for testing, great catch!

I believe I fixed it now:

https://github.com/oilshell/oil/commit/6c9c8a2b3b1d4add7d20752f97c53496458231ee

The problem with the original patch was that `$[5 ** 3]` doesn't contain any parens, but something like `if (5  ** 3 > 0)` does.  I wrote a comment about how that should be harmless, but I was wrong.  (It's a little tricky because the parser can be looking at a `(` token in several function calls) 

I added more tests.   Any other testing is appreciated!
